livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/tts.py

```diff
@@ -68,22 +68,8 @@ async def fetch():
 
                 decoder.close()
 
-        async def decoder_stream():
-            async for data in decoder:
-                results.put_nowait(
-                    tts.SynthesizedAudio(
-                        text=text,
-                        data=rtc.AudioFrame(
-                            data=data,
-                            sample_rate=OPENAI_TTS_SAMPLE_RATE,
-                            num_channels=OPENAI_TTS_CHANNELS,
-                            samples_per_channel=len(data) // 2,
-                        ),
-                    )
-                )
-            results.close()
+        fetch_task = asyncio.create_task(fetch())
+        async for data in decoder:
+            yield tts.SynthesizedAudio(text=text, data=data)
```

Comment:
This isn't streamed to the users. This is waiting for the "download" to finish before yielding anything to the user

Reply:
You could use asyncio.select or __anext__ after a http chunk

Reply:
Hrmm so it isn't waiting for the download to complete. If 4096 bytes come in the decoder starts decoding them. It's possible (and probably) for a few http chunks to come in before the first decoded result but I think that's ok.

I do think decoding should happen on another thread (but maybe just one other thread instead of constantly doing `to_thread`?). I think the code-complexity tradeoff is worth it. Also regardless of internal implementation, I think the decoder api should remain push_chunk + async iterator. 

I'm assuming there will be slower decoders (video?) in the future and having a consistent api probably makes sense, especially when the only downside is a few lines of code. I'm also admittedly a bit scared of debugging issues of users running on cheap hardware decoding longer-than-anticipated audio where the symptoms manifest in non-obvious ways.

I know you've been pretty in favor of sync decoding - are there other reasons that I'm not seeing for your preference?

Reply:
> EDIT: You're creating a new task, I think this should be avoided wdyt?

Hrmm, so I don't have a strong opinion on task creation and whether or not we should try and avoid it. This is moreso a discussion about what should the mp3.StreamDecoder API look like. And _possibly_ a discussion about what should all decoders look like.

I went with push_chunk + a stream decoder mainly because it felt like it communicated that this decoder is stateful. The other option is `async decode_chunk(<some http chunk>) -> rtc.AudioFrame` I'm not totally against but it "feels" stateless even though it's not.